### ISSUE TITLE
Using the exported models 

### ISSUE THREAD

**Issue by danyaljj** [google-research/text-to-text-transfer-transformer#142]
First, thank you for being so helpful + patient with issues/questions. 🙏 

I have a successfully trained/evaluated model and now I am trying to use an exported model created using the construction explained [here](https://github.com/google-research/text-to-text-transfer-transformer#export). 

Here are several efforts:  

(1) Using GC "AI Platform" to deploy models: 

Here I am getting the following error: 
![Screen Shot 2020-03-25 at 7 21 07 PM](https://user-images.githubusercontent.com/2441454/77604371-8eb91000-6ecf-11ea-9890-ded3d1c508fe.png)

```
Create Version failed. Model validation failed: Outer dimension for outputs must be unknown, outer dimension of 'SentenceTokenizer/SentenceTokenizer/SentencepieceDetokenizeOp:0' is 1 For more information on how to export Tensorflow SavedModel, see https://www.tensorflow.org/api_docs/python/tf/saved_model.
```
See the figure below: 

(2) I tried following the code in the Colab: 

```python 
import tensorflow.compat.v1 as tf
tf.reset_default_graph()
TPU_ADDRESS="danielk-tpu-europe-west4-a-v3-8-no4"
saved_model_path="gs://danielk-files/t5-models/squad1_1_race_string_mixture/small/export/1585135433"
sess = tf.Session(TPU_ADDRESS)
meta_graph_def = tf.saved_model.loader.load(sess, ["serve"], saved_model_path)
signature_def = meta_graph_def.signature_def["serving_default"]

def answer(question):
  return sess.run(
      fetches=signature_def.outputs["outputs"].name,
      feed_dict={signature_def.inputs["input"].name: [question]}
  )[0].decode('utf-8')


for question in ["trivia question: where is the google headquarters?",
                 "trivia question: what is the most populous country in the world?",
                 "trivia question: who are the 4 members of the beatles?",
                 "trivia question: how many teeth do humans have?"]:
    print(answer(question))
```

I am getting the following error: 
```
2020-03-26 02:31:17.361676: E tensorflow/core/common_runtime/session.cc:89] Not found: No session factory registered for the given session options: {target: "danielk-tpu-europe-west4-a-v3-8-no4" config: device_count { key: "CPU" value: 1 } device_count { key: "GPU" value: 0 } gpu_options { experimental { } }} Registered factories are {DIRECT_SESSION, GRPC_SESSION}.
Traceback (most recent call last):
  File "predict_single_instance.py", line 5, in <module>
    sess = tf.Session(TPU_ADDRESS)
  File "/home/danielk/anaconda3/envs/env37/lib/python3.7/site-packages/tensorflow_core/python/client/session.py", line 1585, in __init__
    super(Session, self).__init__(target, graph, config=config)
  File "/home/danielk/anaconda3/envs/env37/lib/python3.7/site-packages/tensorflow_core/python/client/session.py", line 699, in __init__
    self._session = tf_session.TF_NewSessionRef(self._graph._c_graph, opts)
tensorflow.python.framework.errors_impl.NotFoundError: No session factory registered for the given session options: {target: "danielk-tpu-europe-west4-a-v3-8-no4" config: device_count { key: "CPU" value: 1 } device_count { key: "GPU" value: 0 } gpu_options { experimental { } }} Registered factories are {DIRECT_SESSION, GRPC_SESSION}.
```

Wondering if you have any suggestions.

**Comment by danyaljj** [comment]
Leaving a small comment here, just in case this falls off your inbox, @adarob.

**Comment by adarob** [comment]
For 1) we'd need to modify the export, but I don't have time to play around with it at the moment. If you are able to figure it out, please help us improve the code for others!

For 2), you're not using the correct address.  A TPU address should look like `grpc://1.2.3.4:8470`.  I believe you can get this address using the `get_master` method of a `TPUClusterResolver` (see https://www.tensorflow.org/api_docs/python/tf/distribute/cluster_resolver/TPUClusterResolver). You can also find it in the GCP console. It's just the internal IP address of the TPU (plus the grpc:// prefix).

Hope this helps!

**Comment by danyaljj** [comment]
On (2), I updated the TPU address, and now the error I am getting is similar with I was getting in (1): 
```
WARNING:tensorflow:From ../predict_single_instance.py:7: load (from tensorflow.python.saved_model.loader_impl) is deprecated and will be removed in a future version.
Instructions for updating:
This function will only be available through the v1 compatibility library as tf.compat.v1.saved_model.loader.load or tf.compat.v1.saved_model.load. There will be a new function for importing SavedModels in Tensorflow 2.0.
Traceback (most recent call last):
  File "../predict_single_instance.py", line 7, in <module>
    meta_graph_def = tf.saved_model.loader.load(sess, ["serve"], saved_model_path)
  File "/home/danielk/anaconda3/envs/env37/lib/python3.7/site-packages/tensorflow_core/python/util/deprecation.py", line 324, in new_func
    return func(*args, **kwargs)
  File "/home/danielk/anaconda3/envs/env37/lib/python3.7/site-packages/tensorflow_core/python/saved_model/loader_impl.py", line 269, in load
    return loader.load(sess, tags, import_scope, **saver_kwargs)
  File "/home/danielk/anaconda3/envs/env37/lib/python3.7/site-packages/tensorflow_core/python/saved_model/loader_impl.py", line 422, in load
    **saver_kwargs)
  File "/home/danielk/anaconda3/envs/env37/lib/python3.7/site-packages/tensorflow_core/python/saved_model/loader_impl.py", line 352, in load_graph
    meta_graph_def, import_scope=import_scope, **saver_kwargs)
  File "/home/danielk/anaconda3/envs/env37/lib/python3.7/site-packages/tensorflow_core/python/training/saver.py", line 1477, in _import_meta_graph_with_return_elements
    **kwargs))
  File "/home/danielk/anaconda3/envs/env37/lib/python3.7/site-packages/tensorflow_core/python/framework/meta_graph.py", line 809, in import_scoped_meta_graph_with_return_elements
    return_elements=return_elements)
  File "/home/danielk/anaconda3/envs/env37/lib/python3.7/site-packages/tensorflow_core/python/util/deprecation.py", line 507, in new_func
    return func(*args, **kwargs)
  File "/home/danielk/anaconda3/envs/env37/lib/python3.7/site-packages/tensorflow_core/python/framework/importer.py", line 405, in import_graph_def
    producer_op_list=producer_op_list)
  File "/home/danielk/anaconda3/envs/env37/lib/python3.7/site-packages/tensorflow_core/python/framework/importer.py", line 501, in _import_graph_def_internal
    graph._c_graph, serialized, options)  # pylint: disable=protected-access
tensorflow.python.framework.errors_impl.NotFoundError: Op type not registered 'SentencepieceOp' in binary running on danielk-tpu-vm-europe-west4-a. Make sure the Op and Kernel are registered in the binary running in this process. Note that if you are loading a saved graph which used ops from tf.contrib, accessing (e.g.) `tf.contrib.resampler` should be done before importing the graph, as contrib ops are lazily registered when the module is first accessed.
```

**Comment by adarob** [comment]
Which version of TF did you launch the TPU with? It probably needs to be
1.15.

**Comment by danyaljj** [comment]
```
t5                       0.3.0
tensorboard              1.15.0
tensorflow               1.15.2
tensorflow-datasets      2.1.0
tensorflow-estimator     1.15.1
tensorflow-metadata      0.21.1
tensorflow-text          1.15.0
```

**Comment by adarob** [comment]
I'm asking a out the TPU, not your VM. You select the version when you
launch it.

**Comment by danyaljj** [comment]
<img width="573" alt="Screen Shot 2020-03-27 at 5 10 37 PM" src="https://user-images.githubusercontent.com/2441454/77809657-e3829500-704d-11ea-96d3-774a0856aa8f.png">

**Comment by danyaljj** [comment]
@adarob do you have any thoughts on this? Stuck here 😢.

**Comment by adarob** [comment]
I see the internal IPs in the table on the page before this one where you list your currently running TPUs.

**Comment by danyaljj** [comment]
hmm ... did follow your previous message.

**Comment by adarob** [comment]
Click the left arrow at the top and you *should* see a table with the internal IP.

**Comment by danyaljj** [comment]
There seems to be a miscommunication. I **did** look up the internal IP via gcloud: 
```
$ gcloud compute tpus list --zone "europe-west4-a"
NAME                                 ZONE            ACCELERATOR_TYPE  NETWORK_ENDPOINTS  NETWORK  RANGE            STATUS
danielk-tpu-europe-west4-a-v3-8-no1  europe-west4-a  v3-8              w.x.y.z:8470    main     x.x.x.x/29    READY
```
But when I used the IP, I get another as mentioned: 
https://github.com/google-research/text-to-text-transfer-transformer/issues/142#issuecomment-605351421

**Comment by danyaljj** [comment]
I tried running a model my CPU machine (a macOS) and I am getting a similar error: 
```
2020-03-31 19:56:13.238885: I tensorflow/core/platform/cpu_feature_guard.cc:142] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 FMA
2020-03-31 19:56:13.255409: I tensorflow/compiler/xla/service/service.cc:168] XLA service 0x7f822a499350 initialized for platform Host (this does not guarantee that XLA will be used). Devices:
2020-03-31 19:56:13.255467: I tensorflow/compiler/xla/service/service.cc:176]   StreamExecutor device (0): Host, Default Version
WARNING:tensorflow:From predict_single_instance.py:10: load (from tensorflow.python.saved_model.loader_impl) is deprecated and will be removed in a future version.
Instructions for updating:
This function will only be available through the v1 compatibility library as tf.compat.v1.saved_model.loader.load or tf.compat.v1.saved_model.load. There will be a new function for importing SavedModels in Tensorflow 2.0.
Traceback (most recent call last):
  File "predict_single_instance.py", line 10, in <module>
    meta_graph_def = tf.saved_model.loader.load(sess, ["serve"], saved_model_path)
  File "/Users/danielk/ideaProjects/t2t-qa/env37/lib/python3.7/site-packages/tensorflow_core/python/util/deprecation.py", line 324, in new_func
    return func(*args, **kwargs)
  File "/Users/danielk/ideaProjects/t2t-qa/env37/lib/python3.7/site-packages/tensorflow_core/python/saved_model/loader_impl.py", line 269, in load
    return loader.load(sess, tags, import_scope, **saver_kwargs)
  File "/Users/danielk/ideaProjects/t2t-qa/env37/lib/python3.7/site-packages/tensorflow_core/python/saved_model/loader_impl.py", line 422, in load
    **saver_kwargs)
  File "/Users/danielk/ideaProjects/t2t-qa/env37/lib/python3.7/site-packages/tensorflow_core/python/saved_model/loader_impl.py", line 352, in load_graph
    meta_graph_def, import_scope=import_scope, **saver_kwargs)
  File "/Users/danielk/ideaProjects/t2t-qa/env37/lib/python3.7/site-packages/tensorflow_core/python/training/saver.py", line 1477, in _import_meta_graph_with_return_elements
    **kwargs))
  File "/Users/danielk/ideaProjects/t2t-qa/env37/lib/python3.7/site-packages/tensorflow_core/python/framework/meta_graph.py", line 809, in import_scoped_meta_graph_with_return_elements
    return_elements=return_elements)
  File "/Users/danielk/ideaProjects/t2t-qa/env37/lib/python3.7/site-packages/tensorflow_core/python/util/deprecation.py", line 507, in new_func
    return func(*args, **kwargs)
  File "/Users/danielk/ideaProjects/t2t-qa/env37/lib/python3.7/site-packages/tensorflow_core/python/framework/importer.py", line 405, in import_graph_def
    producer_op_list=producer_op_list)
  File "/Users/danielk/ideaProjects/t2t-qa/env37/lib/python3.7/site-packages/tensorflow_core/python/framework/importer.py", line 501, in _import_graph_def_internal
    graph._c_graph, serialized, options)  # pylint: disable=protected-access
tensorflow.python.framework.errors_impl.NotFoundError: Op type not registered 'SentencepieceOp' in binary running on danielk0014-2.local. Make sure the Op and Kernel are registered in the binary running in this process. Note that if you are loading a saved graph which used ops from tf.contrib, accessing (e.g.) `tf.contrib.resampler` should be done before importing the graph, as contrib ops are lazily registered when the module is first accessed.
```

**Comment by adarob** [comment]
Can you paste the code you're using to run on CPU?

**Comment by danyaljj** [comment]
Here: 


```
import tensorflow.compat.v1 as tf
tf.reset_default_graph()
saved_model_path="gs://danielk-files/t5-models/race_string_mixture/small/export_d/1585713324/"
sess = tf.Session()
meta_graph_def = tf.saved_model.loader.load(sess, ["serve"], saved_model_path)
signature_def = meta_graph_def.signature_def["serving_default"]

def answer(question):
  return sess.run(
      fetches=signature_def.outputs["outputs"].name,
      feed_dict={signature_def.inputs["input"].name: [question]}
  )[0].decode('utf-8')


for question in ["trivia question: where is the google headquarters?",
                 "trivia question: what is the most populous country in the world?",
                 "trivia question: who are the 4 members of the beatles?",
                 "trivia question: how many teeth do humans have?"]:
    print(answer(question))
```

**Comment by adarob** [comment]
Alright, I'm making a bit of progress on this but need to go for now. I'll update the thread tomorrow.

**Comment by adarob** [comment]
One thing to do is make sure you have tensorflow-text installed on your mac, and maybe imported before calling `tf.saved-model.loader.load`. That is where the SentencepieceOp will come from.

**Comment by danyaljj** [comment]
So following your comment, I added `import tensorflow_text as tf_text` right before `tf.saved_model.loader.load` and it doesn't give me that error anymore on my machine!! 

Now the only remaining issue on this thread is being able to deploy models via GC "AI Platform".

**Comment by adarob** [comment]
Great! I'll update the Colab to add that line. As far as AI Platform is
concerned, we probably need to have them add TF-Text to their binary. I'll
look into it today.

**Comment by danyaljj** [comment]
Thanks, @adarob ! 
I am no fluent w/ TF stuff, but I here is a thought: It might be possible to modify the exported models (or the process of exporting the models) such that a code loading it is forced to load/import `tensorflow_text` as well.

**Comment by adarob** [comment]
It sounds like AI Platform uses TF Serving, which will have tensorflow_text support once v2.2 is released. They are saying it will be 2-4 weeks for that to happen.

**Comment by adarob** [comment]
I'm going to close this out since I've updated the colab in #146. If you're still having an issue with AI Platform after tf-serving is updated, please open a new issue.
